src/cdk/tree/nested-node.ts

```diff
@@ -51,32 +54,33 @@ import {getTreeControlFunctionsMissingError} from './tree-errors';
   providers: [{provide: CdkTreeNode, useExisting: CdkNestedTreeNode}]
 })
 export class CdkNestedTreeNode<T> extends CdkTreeNode<T> implements AfterContentInit, OnDestroy {
+  /** Differ used to find the changes in the data provided by the data source. */
+  private _dataDiffer: IterableDiffer<T>;
+
   /** The children data dataNodes of current node. They will be placed in `CdkTreeNodeOutlet`. */
   protected _children: T[];
 
   /** The children node placeholder. */
   @ContentChildren(CdkTreeNodeOutlet) nodeOutlet: QueryList<CdkTreeNodeOutlet>;
 
   constructor(protected _elementRef: ElementRef,
-              protected _tree: CdkTree<T>) {
+              protected _tree: CdkTree<T>,
+              protected _differs: IterableDiffers) {
     super(_elementRef, _tree);
   }
 
   ngAfterContentInit() {
+    this._dataDiffer = this._differs.find([]).create();
     if (!this._tree.treeControl.getChildren) {
       throw getTreeControlFunctionsMissingError();
     }
     this._tree.treeControl.getChildren(this.data).pipe(takeUntil(this._destroyed))
         .subscribe(result => {
-          if (result && result.length) {
-            // In case when nodeOutlet is not in the DOM when children changes, save it in the node
-            // and add to nodeOutlet when it's available.
-            this._children = result as T[];
-            this._addChildrenNodes();
-          }
+          this._children = result;
+          this._updateChildrenNodes();
         });
     this.nodeOutlet.changes.pipe(takeUntil(this._destroyed))
-        .subscribe((_) => this._addChildrenNodes());
+        .subscribe((_) => this._updateChildrenNodes());
```

Comment:
Not something that was added in this PR, but you can remove the `_` parameter here.